File: workflows/audit/snapshot.py
#!/usr/bin/env python3
"""Usage: snapshot.py   (audit payload JSON on stdin)

Deterministically record one audit run to $HOME/.claude/workspace/history/ and
compute the resolved/new/carried delta against the most recent prior snapshot.
audit.js used to hand this to a general-purpose LLM agent, which walked the file
glob and the finding-by-finding diff token by token and took minutes serially at
the tail of the run. The work is pure bookkeeping (glob, match on file+message,
count, write JSON), so moving it here turns a multi-minute agent into a single
sub-second shell call and makes the delta rule tested code, not prompt adherence.

The workflow script itself cannot touch the filesystem or read env, so the caller
still spawns a shell agent -- but that agent's whole job is now `snapshot.py <
payload`, not reasoning about the diff.

stdin:  JSON {scope, focus, pre_flight, raw_findings[], findings[], skipped[]}
        each raw_findings entry carries at least {file, message}.
stdout: the path of the JSON record written.
exit 0 on success. exit 1 on an unparseable payload (nothing written).

Resolved fields (shell / prior snapshot), added to the record:
  branch        git rev-parse --abbrev-ref HEAD (falls back to "unknown")
  generated_at  UTC ISO-8601
  delta         {resolved, new, carried} counts vs the most recent prior
                audit-*.json, matched on (file, message); first run -> all 0 with
                note "first run".
"""

import glob
import json
import os
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path

HISTORY_DIR = Path(os.path.expanduser("~")) / ".claude" / "workspace" / "history"


def fail(message):
    print(f"Error: {message}", file=sys.stderr)
    sys.exit(1)


def finding_key(f):
    return (f.get("file"), f.get("message"))


def compute_delta(current_raw, prior_raw):
    """resolved = in prior only, new = in current only, carried = in both.

    prior_raw is None when no prior snapshot exists; the caller records a
    "first run" note in that case.
    """
    if prior_raw is None:
        return {"resolved": 0, "new": 0, "carried": 0, "note": "first run"}
    cur = {finding_key(f) for f in current_raw}
    prior = {finding_key(f) for f in prior_raw}
    return {
        "resolved": len(prior - cur),
        "new": len(cur - prior),
        "carried": len(cur & prior),
    }


def latest_prior_raw(history_dir):
    """raw_findings of the most recent prior audit-*.json, or None if none exists.

    Sorted by filename; the name embeds a UTC timestamp so lexical order is
    chronological. A prior file that is unreadable or lacks raw_findings is
    skipped rather than aborting the run.
    """
    priors = sorted(glob.glob(str(history_dir / "audit-*.json")), reverse=True)
    for path in priors:
        try:
            with open(path) as fh:
                data = json.load(fh)
        except (OSError, ValueError):
            continue
        raw = data.get("raw_findings")
        if isinstance(raw, list):
            return raw
    return None


def git_branch():
    try:
        out = subprocess.run(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            capture_output=True,
            text=True,
            timeout=10,
        )
        branch = out.stdout.strip()
        return branch or "unknown"
    except (OSError, subprocess.SubprocessError):
        return "unknown"


def build_record(payload, branch, generated_at, delta):
    record = dict(payload)
    record["branch"] = branch
    record["generated_at"] = generated_at
    record["delta"] = delta
    return record


def main():
    raw_stdin = sys.stdin.read()
    try:
        payload = json.loads(raw_stdin)
    except ValueError as exc:
        fail(f"unparseable payload: {exc}")
    if not isinstance(payload, dict):
        fail("payload must be a JSON object")

    HISTORY_DIR.mkdir(parents=True, exist_ok=True)
    now = datetime.now(timezone.utc)

    prior_raw = latest_prior_raw(HISTORY_DIR)
    current_raw = payload.get("raw_findings") or []
    delta = compute_delta(current_raw, prior_raw)

    record = build_record(
        payload,
        branch=git_branch(),
        generated_at=now.strftime("%Y-%m-%dT%H:%M:%SZ"),
        delta=delta,
    )

    out_path = HISTORY_DIR / f"audit-{now.strftime('%Y-%m-%d-%H%M%S')}.json"
    with open(out_path, "w") as fh:
        json.dump(record, fh, ensure_ascii=False, indent=2)
    print(str(out_path))


if __name__ == "__main__":
    main()

File: workflows/audit/snapshot.ts
#!/usr/bin/env node
/// <reference types="node" />
// Usage: snapshot.ts   (audit payload JSON on stdin)
//
// Record one audit run to $HOME/.claude/history/audit-<YYYY-MM-DD-HHMMSS>.json.
//
// stdin:  JSON {scope, focus, pre_flight, raw_findings[], findings[], skipped[],
//         challenge_ran, verify_ran, tally, needs_context[], zero_reviewer_files[]}
//         each raw_findings entry carries at least {file, message}, plus {id, reviewer,
//         verdict} once the triage pass has run.
//         Every key is copied to the record verbatim; absent keys stay absent.
// stdout: one line of JSON, {path, counts}. counts holds the element count of each
//         array this process serialized.
// exit 0 on success. exit 1 on an unparseable payload (nothing written).
//
// Resolved fields, added to the record:
//   branch        git rev-parse --abbrev-ref HEAD (falls back to "unknown")
//   generated_at  UTC ISO-8601
//
// TypeScript port of the Python audit recorder it replaces.
// Contract: this CLI's own behavior, exercised end to end by
// workflows/audit/tests/snapshot.test.ts against the frozen fixture
// workflows/audit/tests/fixtures/snapshot-cases.json.
import { readFileSync, writeFileSync } from "node:fs";
import { spawnSync } from "node:child_process";
import { homedir } from "node:os";
import { historyPath, isoTimestamp, parsePayload } from "../_lib/cli.ts";
import { isMainModule } from "../_lib/entry-point.ts";

export const COUNTED_ARRAYS = [
  "raw_findings",
  "findings",
  "skipped",
  "needs_context",
  "zero_reviewer_files",
] as const;

/** `git rev-parse --abbrev-ref HEAD`, falling back to "unknown" on a non-zero exit or a
 * spawn error (git missing from PATH), mirroring the Python version's git_branch.
 * Runs with a 10s timeout, same as the Python version's subprocess.run(timeout=10). */
export function gitBranch(): string {
  const result = spawnSync("git", ["rev-parse", "--abbrev-ref", "HEAD"], {
    encoding: "utf8",
    timeout: 10_000,
  });
  if (result.error || result.status !== 0) return "unknown";
  const branch = result.stdout.trim();
  return branch || "unknown";
}

/** An absent key counts 0 rather than being omitted, so the caller reads the same key set
 * every run and a dropped array is a count mismatch instead of a missing field. Mirrors
 * the Python version's counted_arrays. */
export function countedArrays(record: Record<string, unknown>): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const key of COUNTED_ARRAYS) {
    const value = record[key];
    counts[key] = Array.isArray(value) ? value.length : 0;
  }
  return counts;
}

/** The record written to disk: the payload's own keys in order, then `branch`, then
 * `generated_at`. Mirrors the Python version's build_record. */
export function buildRecord(
  payload: Record<string, unknown>,
  branch: string,
  generatedAt: string,
): Record<string, unknown> {
  return { ...payload, branch, generated_at: generatedAt };
}

export function main(): number {
  const raw = readFileSync(0, "utf8");
  const { payload, message } = parsePayload(raw);
  if (payload === null) {
    process.stderr.write(`${message}\n`);
    return 1;
  }

  // One `now` feeds both the record's generated_at and the record file's own timestamp
  // component, so the two always agree, mirroring the Python version's single `now`.
  const now = new Date();
  const generatedAt = isoTimestamp(now);
  const stamp = generatedAt.replace(/:/g, "").replace("T", "-").replace("Z", "");

  const record = buildRecord(payload, gitBranch(), generatedAt);

  // historyPath ensures $HOME/.claude/history exists and returns the full path, the same
  // flow workflows/build/record.ts uses for its own history file -- the only difference is
  // this recorder's `name` is minted per run instead of a fixed .jsonl name.
  const outPath = historyPath(homedir(), `audit-${stamp}.json`);
  writeFileSync(outPath, `${JSON.stringify(record, null, 2)}\n`);

  const output = { path: outPath, counts: countedArrays(record) };
  process.stdout.write(`${JSON.stringify(output)}\n`);
  return 0;
}

if (isMainModule(import.meta.url)) {
  process.exit(main());
}
